File: db4on/core/src/core/com/db4o/internal/PrimitiveFieldHandler.cs
namespace com.db4o.@internal
{
	/// <exclude></exclude>
	public class PrimitiveFieldHandler : com.db4o.@internal.ClassMetadata
	{
		public readonly com.db4o.@internal.TypeHandler4 i_handler;

		internal PrimitiveFieldHandler(com.db4o.@internal.ObjectContainerBase a_stream, com.db4o.@internal.TypeHandler4
			 a_handler) : base(a_stream, a_handler.ClassReflector())
		{
			i_fields = com.db4o.@internal.FieldMetadata.EMPTY_ARRAY;
			i_handler = a_handler;
		}

		internal override void ActivateFields(com.db4o.@internal.Transaction a_trans, object
			 a_object, int a_depth)
		{
		}

		internal sealed override void AddToIndex(com.db4o.@internal.LocalObjectContainer 
			a_stream, com.db4o.@internal.Transaction a_trans, int a_id)
		{
		}

		internal override bool AllowsQueries()
		{
			return false;
		}

		internal override void CacheDirty(com.db4o.foundation.Collection4 col)
		{
		}

		public override bool CanHold(com.db4o.reflect.ReflectClass claxx)
		{
			return i_handler.CanHold(claxx);
		}

		public override com.db4o.reflect.ReflectClass ClassReflector()
		{
			return i_handler.ClassReflector();
		}

		public override void DeleteEmbedded(com.db4o.@internal.marshall.MarshallerFamily 
			mf, com.db4o.@internal.StatefulBuffer a_bytes)
		{
			if (mf._primitive.UseNormalClassRead())
			{
				base.DeleteEmbedded(mf, a_bytes);
				return;
			}
		}

		public override void DeleteEmbedded1(com.db4o.@internal.marshall.MarshallerFamily
			 mf, com.db4o.@internal.StatefulBuffer a_bytes, int a_id)
		{
			if (i_handler is com.db4o.@internal.handlers.ArrayHandler)
			{
				com.db4o.@internal.handlers.ArrayHandler ya = (com.db4o.@internal.handlers.ArrayHandler
					)i_handler;
				if (ya.i_isPrimitive)
				{
					ya.DeletePrimitiveEmbedded(a_bytes, this);
					a_bytes.SlotDelete();
					return;
				}
			}
			if (i_handler is com.db4o.@internal.UntypedFieldHandler)
			{
				a_bytes.IncrementOffset(i_handler.LinkLength());
			}
			else
			{
				i_handler.DeleteEmbedded(mf, a_bytes);
			}
			Free(a_bytes, a_id);
		}

		internal override void DeleteMembers(com.db4o.@internal.marshall.MarshallerFamily
			 mf, com.db4o.@internal.marshall.ObjectHeaderAttributes attributes, com.db4o.@internal.StatefulBuffer
			 a_bytes, int a_type, bool isUpdate)
		{
			if (a_type == com.db4o.@internal.Const4.TYPE_ARRAY)
			{
				new com.db4o.@internal.handlers.ArrayHandler(a_bytes.GetStream(), this, true).DeletePrimitiveEmbedded
					(a_bytes, this);
			}
			else
			{
				if (a_type == com.db4o.@internal.Const4.TYPE_NARRAY)
				{
					new com.db4o.@internal.handlers.MultidimensionalArrayHandler(a_bytes.GetStream(), 
						this, true).DeletePrimitiveEmbedded(a_bytes, this);
				}
			}
		}

		internal void Free(com.db4o.@internal.Transaction a_trans, int a_id, int a_address
			, int a_length)
		{
			a_trans.SlotFreePointerOnCommit(a_id, a_address, a_length);
		}

		internal void Free(com.db4o.@internal.StatefulBuffer a_bytes, int a_id)
		{
			a_bytes.GetTransaction().SlotFreePointerOnCommit(a_id, a_bytes.GetAddress(), a_bytes
				.GetLength());
		}

		public override bool HasIndex()
		{
			return false;
		}

		internal override object Instantiate(com.db4o.@internal.ObjectReference a_yapObject
			, object a_object, com.db4o.@internal.marshall.MarshallerFamily mf, com.db4o.@internal.marshall.ObjectHeaderAttributes
			 attributes, com.db4o.@internal.StatefulBuffer a_bytes, bool a_addToIDTree)
		{
			if (a_object == null)
			{
				try
				{
					a_object = i_handler.Read(mf, a_bytes, true);
				}
				catch (com.db4o.CorruptionException)
				{
					return null;
				}
				a_yapObject.SetObjectWeak(a_bytes.GetStream(), a_object);
			}
			a_yapObject.SetStateClean();
			return a_object;
		}

		internal override object InstantiateTransient(com.db4o.@internal.ObjectReference 
			a_yapObject, object a_object, com.db4o.@internal.marshall.MarshallerFamily mf, com.db4o.@internal.marshall.ObjectHeaderAttributes
			 attributes, com.db4o.@internal.StatefulBuffer a_bytes)
		{
			try
			{
				return i_handler.Read(mf, a_bytes, true);
			}
			catch (com.db4o.CorruptionException)
			{
				return null;
			}
		}

		internal override void InstantiateFields(com.db4o.@internal.ObjectReference a_yapObject
			, object a_onObject, com.db4o.@internal.marshall.MarshallerFamily mf, com.db4o.@internal.marshall.ObjectHeaderAttributes
			 attributes, com.db4o.@internal.StatefulBuffer a_bytes)
		{
			object obj = null;
			try
			{
				obj = i_handler.Read(mf, a_bytes, true);
			}
			catch (com.db4o.CorruptionException)
			{
			}
			if (obj != null)
			{
				i_handler.CopyValue(obj, a_onObject);
			}
		}

		public override bool IsArray()
		{
			return i_id == com.db4o.@internal.HandlerRegistry.ANY_ARRAY_ID || i_id == com.db4o.@internal.HandlerRegistry
				.ANY_ARRAY_N_ID;
		}

		public override bool IsPrimitive()
		{
			return true;
		}

		public override int IsSecondClass()
		{
			return com.db4o.@internal.Const4.UNKNOWN;
		}

		public override bool IsStrongTyped()
		{
			return false;
		}

		public override void CalculateLengths(com.db4o.@internal.Transaction trans, com.db4o.@internal.marshall.ObjectHeaderAttributes
			 header, bool topLevel, object obj, bool withIndirection)
		{
			i_handler.CalculateLengths(trans, header, topLevel, obj, withIndirection);
		}

		public override com.db4o.@internal.Comparable4 PrepareComparison(object a_constraint
			)
		{
			i_handler.PrepareComparison(a_constraint);
			return i_handler;
		}

		public sealed override com.db4o.reflect.ReflectClass PrimitiveClassReflector()
		{
			return i_handler.PrimitiveClassReflector();
		}

		public override object Read(com.db4o.@internal.marshall.MarshallerFamily mf, com.db4o.@internal.StatefulBuffer
			 a_bytes, bool redirect)
		{
			if (mf._primitive.UseNormalClassRead())
			{
				return base.Read(mf, a_bytes, redirect);
			}
			return i_handler.Read(mf, a_bytes, false);
		}

		public override com.db4o.@internal.TypeHandler4 ReadArrayHandler(com.db4o.@internal.Transaction
			 a_trans, com.db4o.@internal.marshall.MarshallerFamily mf, com.db4o.@internal.Buffer[]
			 a_bytes)
		{
			if (IsArray())
			{
				return i_handler;
			}
			return null;
		}

		public override object ReadQuery(com.db4o.@internal.Transaction trans, com.db4o.@internal.marshall.MarshallerFamily
			 mf, bool withRedirection, com.db4o.@internal.Buffer reader, bool toArray)
		{
			if (mf._primitive.UseNormalClassRead())
			{
				return base.ReadQuery(trans, mf, withRedirection, reader, toArray);
			}
			return i_handler.ReadQuery(trans, mf, withRedirection, reader, toArray);
		}

		public override com.db4o.@internal.query.processor.QCandidate ReadSubCandidate(com.db4o.@internal.marshall.MarshallerFamily
			 mf, com.db4o.@internal.Buffer reader, com.db4o.@internal.query.processor.QCandidates
			 candidates, bool withIndirection)
		{
			return i_handler.ReadSubCandidate(mf, reader, candidates, withIndirection);
		}

		internal override void RemoveFromIndex(com.db4o.@internal.Transaction ta, int id)
		{
		}

		public override bool SupportsIndex()
		{
			return true;
		}

		public sealed override bool WriteObjectBegin()
		{
			return false;
		}

		public override object WriteNew(com.db4o.@internal.marshall.MarshallerFamily mf, 
			object a_object, bool topLevel, com.db4o.@internal.StatefulBuffer a_bytes, bool 
			withIndirection, bool restoreLinkOffset)
		{
			mf._primitive.WriteNew(a_bytes.GetTransaction(), this, a_object, topLevel, a_bytes
				, withIndirection, restoreLinkOffset);
			return a_object;
		}

		public override string ToString()
		{
			return "Wraps " + i_handler.ToString() + " in YapClassPrimitive";
		}

		public override void Defrag(com.db4o.@internal.marshall.MarshallerFamily mf, com.db4o.@internal.ReaderPair
			 readers, bool redirect)
		{
			if (mf._primitive.UseNormalClassRead())
			{
				base.Defrag(mf, readers, redirect);
			}
			else
			{
				i_handler.Defrag(mf, readers, false);
			}
		}
	}
}

File: db4oj/core/src/com/db4o/internal/PrimitiveFieldHandler.java
/* Copyright (C) 2004 - 2008  db4objects Inc.   http://www.db4o.com */

package com.db4o.internal;

import com.db4o.ext.*;
import com.db4o.foundation.*;
import com.db4o.internal.activation.*;
import com.db4o.internal.delete.*;
import com.db4o.internal.fieldhandlers.*;
import com.db4o.internal.handlers.*;
import com.db4o.internal.handlers.array.*;
import com.db4o.internal.marshall.*;
import com.db4o.marshall.*;
import com.db4o.reflect.*;
import com.db4o.typehandlers.*;


/**
 * @exclude
 */
public class PrimitiveFieldHandler extends ClassMetadata implements FieldHandler, VersionedTypeHandler{
    
    private static final int HASHCODE_FOR_NULL = 283636383;
    
    private final TypeHandler4 _handler;
    
    public PrimitiveFieldHandler(ObjectContainerBase container, TypeHandler4 handler, int handlerID, ReflectClass classReflector) {
    	super(container, classReflector);
        _aspects = FieldMetadata.EMPTY_ARRAY;
        _handler = handler;
        _id = handlerID;
    }
    
    public PrimitiveFieldHandler(){
        super(null, null);
        _handler = null;
    }

    public void activateFields(Transaction trans, Object obj, ActivationDepth depth) {
        // Override
        // do nothing
    }

    final void addToIndex(Transaction trans, int id) {
        // Override
        // Primitive Indices will be created later.
    }

    boolean allowsQueries() {
        return false;
    }

    void cacheDirty(Collection4 col) {
        // do nothing
    }
    
    public boolean descendOnCascadingActivation() {
        return false;
    }

    public void delete(DeleteContext context) throws Db4oIOException {
    	if(context.isLegacyHandlerVersion()){
    		context.readInt();
    		context.defragmentRecommended();
    	}
    }

    void deleteMembers(MarshallerFamily mf, ObjectHeaderAttributes attributes, StatefulBuffer a_bytes, int a_type, boolean isUpdate) {
        if (a_type == Const4.TYPE_ARRAY) {
            new ArrayHandler(this, true).deletePrimitiveEmbedded(a_bytes, this);
        } else if (a_type == Const4.TYPE_NARRAY) {
            new MultidimensionalArrayHandler(this, true).deletePrimitiveEmbedded(a_bytes, this);
        }
    }
    
    void deleteMembers(DeleteContextImpl context, int a_type, boolean isUpdate) {
        if (a_type == Const4.TYPE_ARRAY) {
            new ArrayHandler(this, true).deletePrimitiveEmbedded((StatefulBuffer) context.buffer(), this);
        } else if (a_type == Const4.TYPE_NARRAY) {
            new MultidimensionalArrayHandler(this, true).deletePrimitiveEmbedded((StatefulBuffer) context.buffer(), this);
        }
    }
    
	final void free(StatefulBuffer a_bytes, int a_id) {
          a_bytes.transaction().slotFreePointerOnCommit(a_id, a_bytes.slot());
	}
    
	public boolean hasClassIndex() {
	    return false;
	}

    public Object instantiate(UnmarshallingContext context) {
        Object obj = context.persistentObject();
        if (obj == null) {
            obj = context.read(_handler);
            context.setObjectWeak(obj);
        }
        context.setStateClean();
        return obj;
    }
    
    public Object instantiateTransient(UnmarshallingContext context) {
        return _handler.read(context);
    }

    Object instantiateFields(UnmarshallingContext context) {
        Object obj = context.read(_handler);
        if (obj == null  ||  ! (_handler instanceof DateHandler)) {
            return obj;
        }
        final Object existing = context.persistentObject();
		Object newValue = dateHandler().copyValue(obj, existing);
		
        // FIXME: It should not be necessary to set persistentObject here
        context.persistentObject(newValue);
        
		return newValue;
    }

	private DateHandler dateHandler() {
		return ((DateHandler)_handler);
	}

    public boolean isArray() {
        return _id == Handlers4.ANY_ARRAY_ID || _id == Handlers4.ANY_ARRAY_N_ID;
    }
    
    public boolean isPrimitive(){
        return true;
    }
    
	public boolean isStrongTyped(){
		return false;
	}
    
    public PreparedComparison prepareComparison(Context context, Object source) {
    	return _handler.prepareComparison(context, source);
    }
    
    public TypeHandler4 readCandidateHandler(QueryingReadContext context) {
        if (isArray()) {
            return _handler;
        }
        return null;
    }
    
    public ObjectID readObjectID(InternalReadContext context){
        if(_handler instanceof ClassMetadata){
            return ((ClassMetadata)_handler).readObjectID(context);
        }
        if(_handler instanceof ArrayHandler){
            // TODO: Here we should theoretically read through the array and collect candidates.
            // The respective construct is wild: "Contains query through an array in an array."
            // Ignore for now.
            return ObjectID.IGNORE;
        }
        return ObjectID.NOT_POSSIBLE;
    }
    
    void removeFromIndex(Transaction ta, int id) {
        // do nothing
    }
    
    public final boolean writeObjectBegin() {
        return false;
    }
    
    public String toString(){
        return "Wraps " + _handler.toString() + " in YapClassPrimitive";
    }

    public void defragment(DefragmentContext context) {
    	_handler.defragment(context);
    }
    
    public Object wrapWithTransactionContext(Transaction transaction, Object value) {
        return value;
    }
    
    public Object read(ReadContext context) {
        return _handler.read(context);
    }
    
    public void write(WriteContext context, Object obj) {
        _handler.write(context, obj);
    }
    
    public TypeHandler4 typeHandler(){
        return _handler;
    }
    
    public TypeHandler4 delegateTypeHandler(){
        return _handler;
    }
    
    public TypeHandler4 unversionedTemplate() {
        return new PrimitiveFieldHandler(null, null, 0, null);
    }
    
    public boolean equals(Object obj) {
        if(! (obj instanceof PrimitiveFieldHandler)){
            return false;
        }
        PrimitiveFieldHandler other = (PrimitiveFieldHandler) obj;
        if(_handler == null){
            return other._handler == null;
        }
        return _handler.equals(other._handler);
    }
    
    public int hashCode() {
        if(_handler == null){
            return HASHCODE_FOR_NULL;
        }
        return _handler.hashCode();
    }
    
    public Object deepClone(Object context) {
        TypeHandlerCloneContext typeHandlerCloneContext = (TypeHandlerCloneContext) context;
        PrimitiveFieldHandler original = (PrimitiveFieldHandler) typeHandlerCloneContext.original;
        TypeHandler4 delegateTypeHandler = typeHandlerCloneContext.correctHandlerVersion(original.delegateTypeHandler());
        return new PrimitiveFieldHandler(original.container(), delegateTypeHandler, original._id, original.classReflector());
    }
    
    public boolean isSecondClass() {
    	return isSecondClass(_handler);
    }


}
